src/auth/digest/Config.cc

```diff
@@ -151,85 +156,61 @@ authenticateDigestNonceNew(void)
     static xuniform_int_distribution<uint32_t> newRandomData;
 
     /* create a new nonce */
-    digest_nonce_h *newnonce = new digest_nonce_h;
-    newnonce->noncedata.creationtime = current_time.tv_sec;
-    newnonce->noncedata.randomdata = newRandomData(mt);
-
-    authDigestNonceEncode(newnonce);
+    digest_nonce_h::Pointer newnonce = new digest_nonce_h;
+    newnonce->encode(newRandomData(mt));
 
     // ensure temporal uniqueness by checking for existing nonce
-    while (authenticateDigestNonceFindNonce((char const *) (newnonce->key))) {
+    while (NonceCache().find(newnonce->hexKey) != NonceCache().end()) {
         /* create a new nonce */
-        newnonce->noncedata.randomdata = newRandomData(mt);
-        authDigestNonceEncode(newnonce);
+        newnonce->encode(newRandomData(mt));
     }
 
-    hash_join(digest_nonce_cache, newnonce);
-    /* the cache's link */
-    authDigestNonceLink(newnonce);
-    newnonce->flags.incache = true;
+    newnonce->cache();
     debugs(29, 5, "created nonce " << newnonce << " at " << newnonce->noncedata.creationtime);
     return newnonce;
 }
 
 static void
 authenticateDigestNonceSetup(void)
 {
-    if (!digest_nonce_cache) {
-        digest_nonce_cache = hash_create((HASHCMP *) strcmp, 7921, hash_string);
-        assert(digest_nonce_cache);
-        eventAdd("Digest nonce cache maintenance", authenticateDigestNonceCacheCleanup, nullptr, static_cast<Auth::Digest::Config*>(Auth::SchemeConfig::Find("digest"))->nonceGCInterval, 1);
-    }
+    eventAdd("Digest nonce cache maintenance", authenticateDigestNonceCacheCleanup, nullptr, static_cast<Auth::Digest::Config*>(Auth::SchemeConfig::Find("digest"))->nonceGCInterval, 1);
 }
 
 void
 authenticateDigestNonceShutdown(void)
 {
-    /*
-     * We empty the cache of any nonces left in there.
-     */
-    digest_nonce_h *nonce;
-
-    if (digest_nonce_cache) {
-        debugs(29, 2, "Shutting down nonce cache");
-        hash_first(digest_nonce_cache);
-
-        while ((nonce = ((digest_nonce_h *) hash_next(digest_nonce_cache)))) {
-            assert(nonce->flags.incache);
-            authDigestNoncePurge(nonce);
-        }
+    // empty the cache of any nonces left in there
+    for (const auto &entry: NonceCache()) {
+        entry.second->purge();
     }
```

Comment:
Modifying the container we are iterating using a for-range loop is a UB -- the loop does not compute and store the next iterator value before we invalidate the current iterator.

Another, more subtle problem here (or in the purge() method) is that purge() is very forgiving -- it does nothing if the entry is not cached. It _should_ be cached, but if that implicit invariant is violated, we will loop forever here.

The flags.incache hack complicates things quite a bit, but I suspect the best solution here is either 

* a while-not-empty loop that purges the begin() entry and asserts that purge() was successful; that assertion can be implemented without relying on purge() internals by comparing the cache size() before and after purge()
* a for-range loop that clears flags.incache, followed by a single NonceCache.clear() call; this somewhat invasive option can be viewed as a performance optimization for large caches

If these caches may be large, I would go for the second option, especially since flags.incache itself is essentially an optimization (AFAICT).

Reply:
Alternatively, consider doing nothing in this shutdown method (other than asserting that we are shutting down, perhaps). What is the point of cleaning up a _temporary cache_ in a to-be-terminated program? Such a cache is essentially an optional optimization that we should be able to ignore on shutdown, right?

Reply:
These are std::list - "Adding, removing and moving the elements within the list or across several lists does not invalidate the iterators or references. An iterator is invalidated only when the corresponding element is deleted. "
Which specifically means that ranged for are safely defined behaviour.

Reply:
> An iterator is invalidated only when the corresponding element is deleted

... which is exactly what this code is doing.

>  Which specifically means that ranged for are safely defined behaviour.

... unless that loop deletes the current element.

Reply:
> These are std::list

This loop iterates std::unordered_map. However, in this digest_nonce_h::purge()/NonceCache().erase() context, std::unordered_map has the same iterator invalidation problem as std::list would have had -- the erase() method invalidates the current iterator.

Reply:
https://en.cppreference.com/w/cpp/container/unordered_map/erase
"The order of the elements that are not erased is preserved. (This makes it possible to erase individual elements while iterating through the container.) "

Reply:
> The order of the elements that are not erased is preserved. (This makes it possible to erase individual elements while iterating through the container.)

It is, indeed, possible to erase individual elements while iterating through std::unordered_map. Nothing in this thread contradicts that fact! However, we still have to avoid using invalidated iterators when erasing-while-iterating. There is no guarantee that every range-based `for` loop implementation encountered by Squid will avoid that UB. Moreover, a well-known _natural_ [implementation](https://en.cppreference.com/w/cpp/language/range-for) triggers that UB when it increments `__begin` iterator invalidated by the `loop-statement`:

```C++
for (auto __begin = begin-expr, __end = end-expr; __begin != __end; ++__begin) {
    range-declaration = *__begin;
    loop-statement
}
```

In my [test](https://godbolt.org/z/zz6G5GcoT), GCC increments the iterator invalidated by the `loop-statement` in a range-based `for` loop, matching the above "natural implementation" of that loop (AFAICT):

```asm
    call std::__detail::_Node_iterator<...>::operator*() const
    ...
    call std::unordered_map<...>::erase(int const&)
    ...
    call std::__detail::_Node_iterator<...>::operator++()
```